File: paseo-plugin/host/recovery.mjs
import { createHash } from "node:crypto";

export const PROVIDER_ATTEMPTS = 3;
export const MALFORMED_REGENS = 1;
export const PROCESS_RESTARTS = 1;
export const AGGREGATE_RECOVERY = 6;
export const LOOP_HISTORY = 20;
export const LOOP_WARN_AT = 3;
export const LOOP_STOP_AT = 6;
export const CIRCUIT_FAILURES = 5;
export const CIRCUIT_MIN_REQUESTS = 2;
export const CIRCUIT_COOLDOWN_MS = 60_000;
export const BACKOFF_CAPS_MS = Object.freeze([5_000, 10_000]);

export function failureText(error) {
  if (error == null) return "";
  const stderr = String(error.stderr ?? "").trim();
  if (stderr) return stderr;
  if (error instanceof Error) return error.message;
  return String(error);
}

export function classifyFailure(error) {
  if (error?.name === 'AbortError' || ['SIGTERM', 'SIGINT'].includes(error?.signal)) return 'cancelled';
  const seen = new Set();
  for (let item = error; item && !seen.has(item); item = item.cause) {
    seen.add(item);
    if (item.failureClass) return item.failureClass;
    const status = item.status ?? item.statusCode ?? item.response?.status;
    if ([400, 401, 403, 404, 422].includes(status)) return "permanent";
    if ([408, 429, 500, 502, 503, 504].includes(status)) return "transient";
  }
  const s = [...seen].map(failureText).join(" | ").toLowerCase();
  if (/degeneration|repeated-action loop|loop persisted/.test(s)) return "degeneration";
  if (
    /set brave_api_key|exa_api_key|missing search credentials|401\b|403\b|\b400\b|invalid api key|authentication|insufficient_quota|quota exceeded|unsupported model|invalid_request|invalid-argument|invalid_argument|badrequesterror|bad request|maximum prompt|context length|context window|too many tokens|token limit/.test(s)
    && !/timeout|502|503|429/.test(s)
  ) {
    return "permanent";
  }
  if (
    /\b429\b|rate.?limit|\b50[0234]\b|econnreset|etimedout|enotfound|readtimeout|timeout after|internal error during token|service unavailable|temporarily unavailable|circuit open/.test(s)
  ) {
    return "transient";
  }
  if (/produced no output|produced no json|output is not json|empty model response|question is empty/.test(s)) {
    return "invalid_output";
  }
  return "process";
}

export function summarizeFailure(error) {
  const text = failureText(error).replace(/\s+/g, " ").trim();
  const timeout = text.match(/timeout after ([\d.]+)\s*seconds/i);
  if (timeout) return `ReadTimeout at the provider's ${timeout[1]}-second timeout`;
  const xai = text.match(/"error"\s*:\s*"((?:\\.|[^"\\])*)"/);
  if (xai) return xai[1].replace(/\\"/g, '"');
  if (/internal error during token|\b500\b/.test(text) && !/\b400\b/.test(text)) return "HTTP 500";
  if (/readtimeout/i.test(text)) return "ReadTimeout";
  const http = text.match(/\bHTTP\s*([45]\d\d)\b/i);
  if (http) return `HTTP ${http[1]}`;
  if (text.length <= 240) return text;
  return text.slice(-240);
}

export function createLedger() {
  return {
    providerAttempts: 0,
    regenerations: 0,
    processRestarts: 0,
    recoveryActions: 0,
  };
}

export function canRecover(ledger) {
  return (ledger?.recoveryActions ?? 0) < AGGREGATE_RECOVERY;
}

export function backoffMs(failedAttempt, random = Math.random) {
  const cap = BACKOFF_CAPS_MS[Math.min(Math.max(failedAttempt, 1), BACKOFF_CAPS_MS.length) - 1];
  return Math.floor((random() ?? 0) * cap);
}

export function createCircuit({
  now = Date.now,
  openAfter = CIRCUIT_FAILURES,
  minRequests = CIRCUIT_MIN_REQUESTS,
  cooldownMs = CIRCUIT_COOLDOWN_MS,
  store,
} = {}) {
  const scopes = new Map(store?.list("circuit") ?? []);
  const save = (scope, item) => store?.put("circuit", String(scope ?? "default"), item);
  function bucket(scope) {
    const key = String(scope ?? "default");
    if (!scopes.has(key)) {
      scopes.set(key, {
        consecutive: 0,
        requests: 0,
        state: "closed",
        openedAt: 0,
        probe: false,
      });
    }
    return scopes.get(key);
  }
  function maybeHalfOpen(item) {
    if (item.state === "open" && now() - item.openedAt >= cooldownMs) {
      item.state = "half-open";
      item.probe = true;
    }
  }
  return {
    isOpen(scope) {
      const item = bucket(scope);
      maybeHalfOpen(item);
      if (item.state === "closed") return false;
      if (item.state === "half-open" && item.probe) {
        item.probe = false;
        save(scope, item);
        return false;
      }
      return item.state !== "closed";
    },
    recordSuccess(scope) {
      const item = bucket(scope);
      item.consecutive = 0;
      item.requests = 0;
      item.state = "closed";
      item.probe = false;
      save(scope, item);
    },
    recordFailure(scope, cls, { newRequest = false } = {}) {
      if (cls !== "transient") return;
      const item = bucket(scope);
      if (newRequest) item.requests += 1;
      item.consecutive += 1;
      if (item.state === "half-open") {
        item.state = "open";
        item.openedAt = now();
        item.probe = false;
        save(scope, item);
        return;
      }
      if (item.consecutive >= openAfter && item.requests >= minRequests) {
        item.state = "open";
        item.openedAt = now();
      }
      save(scope, item);
    },
    snapshot(scope) {
      const item = bucket(scope);
      maybeHalfOpen(item);
      return { ...item };
    },
  };
}

export async function retryProviderOperation(fn, {
  ledger = createLedger(),
  circuit,
  scope = "default",
  sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms)),
  random = Math.random,
} = {}) {
  const attempts = [];
  if (circuit?.isOpen(scope)) {
    const error = new Error("provider circuit open");
    error.failureClass = "transient";
    error.circuitOpen = true;
    error.attempts = attempts;
    throw error;
  }
  let requestCounted = false;
  for (let n = 1; n <= PROVIDER_ATTEMPTS; n++) {
    try {
      const result = await fn();
      circuit?.recordSuccess(scope);
      ledger.providerAttempts = n;
      return result;
    } catch (error) {
      if (error.supervised) throw error;
      const cls = error.failureClass ?? classifyFailure(error);
      attempts.push(summarizeFailure(error));
      if (!requestCounted) {
        circuit?.recordFailure(scope, cls, { newRequest: true });
        requestCounted = true;
      } else {
        circuit?.recordFailure(scope, cls, { newRequest: false });
      }
      error.failureClass = cls;
      error.attempts = attempts;
      if (cls === "permanent" || cls === "degeneration") throw error;
      if (cls === "process") throw error;
      if (cls === "invalid_output") {
        if (ledger.regenerations >= MALFORMED_REGENS) {
          error.exhausted = true;
          throw error;
        }
        ledger.regenerations += 1;
      }
      const more = n < PROVIDER_ATTEMPTS && canRecover(ledger);
      if (!more) {
        error.exhausted = cls === "transient" || cls === "invalid_output";
        throw error;
      }
      ledger.recoveryActions += 1;
      ledger.providerAttempts = n;
      const after = error.retryAfter ?? error.response?.headers?.get?.("retry-after");
      const retryAfterMs = after == null ? 0 : Number.isFinite(Number(after)) ? Number(after) * 1000 : Math.max(0, Date.parse(after) - Date.now());
      await sleep(Math.max(backoffMs(n, random), retryAfterMs || 0));
      if (circuit?.isOpen(scope)) { error.circuitOpen = true; throw error; }
    }
  }
  const error = new Error(attempts.join(" → ") || "provider recovery exhausted");
  error.failureClass = "transient";
  error.attempts = attempts;
  error.exhausted = true;
  throw error;
}

export function operationKey(cwd, kind, identity) {
  return createHash("sha256").update(`${cwd}\n${kind}\n${identity}`).digest("hex");
}

export function parseResearcherResult(stderr) {
  const text = String(stderr ?? "");
  const lines = text.split(/\r?\n/).reverse();
  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed.startsWith("RESEARCHER_RESULT ")) continue;
    try {
      return JSON.parse(trimmed.slice("RESEARCHER_RESULT ".length));
    } catch {
      return null;
    }
  }
  return null;
}

export function formatSpentProviderWake({
  role = "research",
  attempts = [],
  details = "",
  headSha = "",
  worktree = "",
} = {}) {
  const n = Math.max(attempts.length, 1);
  const unit = n === 1 ? "attempt" : "attempts";
  const chain = attempts.length ? attempts.join(" → ") : "provider failure";
  if (role === "review") {
    const where = [
      headSha ? `at ${headSha}` : "",
      worktree ? `in ${worktree}` : "",
    ].filter(Boolean).join(" ");
    return [
      `Review could not complete after ${n} provider ${unit}${attempts.length ? `: ${chain}` : ""}.`,
      `The submitted change is preserved${where ? ` ${where}` : ""}; it has not been landed.`,
      "No review verdict was accepted, and the correction allowance was not consumed.",
      "Automatic recovery is exhausted.",
      details ? `Details: ${details}.` : null,
    ].filter(Boolean).join(" ");
  }
  return [
    `Research could not complete after ${n} provider ${unit}: ${chain}.`,
    "Automatic recovery is exhausted.",
    "No answer was accepted.",
    details ? `Details: ${details}.` : null,
  ].filter(Boolean).join(" ");
}

export function formatPermanentWake({ role = "research", summary, details } = {}) {
  const who = role === "review" ? "Review" : "Research";
  return [
    `${who} could not complete: ${summary}.`,
    "Automatic recovery was not attempted.",
    details ? `Details: ${details}.` : null,
  ].filter(Boolean).join(" ");
}

export function formatDegenerationWake({
  role = "implementer",
  action = "repeated tool calls",
  occurrences = LOOP_STOP_AT,
  worktree = "",
  details = "",
} = {}) {
  const who = role === "research" ? "Research" : "Implementer";
  return [
    `${who} stopped after a repeated-action loop persisted despite a warning: ${action}, unchanged results, ${occurrences} occurrences.`,
    worktree
      ? `Work is preserved in ${worktree}; it has not been landed.`
      : "Work is preserved; it has not been landed.",
    "Automatic restart was not attempted.",
    details ? `Details: ${details}.` : null,
  ].filter(Boolean).join(" ");
}

export function formatLivenessWake({ role = "research", operation = "a provider request", details = "" } = {}) {
  const who = role === "review" ? "Review" : "Research";
  return [
    `${who} liveness is uncertain while waiting on ${operation}.`,
    "The process has not been confirmed stopped, and no replacement was started.",
    "Operator inspection or cancellation is needed.",
    details ? `Details: ${details}.` : null,
  ].filter(Boolean).join(" ");
}

export function formatCircuitWake({ role = "research", details = "" } = {}) {
  const who = role === "review" ? "Review" : "Research";
  return [
    `${who} is blocked: the provider is failing.`,
    "Automatic recovery is paused for 60 seconds.",
    details ? `Details: ${details}.` : null,
  ].filter(Boolean).join(" ");
}

export function formatRecoveryWake(error, { role = "research", details = "", headSha = "", worktree = "" } = {}) {
  const cls = error?.failureClass ?? classifyFailure(error);
  if (cls === 'cancelled') return `${role === 'review' ? 'Review' : role === 'implementer' ? 'Implementer' : 'Research'} cancelled. Existing work and request evidence are preserved; no replacement was started.${details ? ` Details: ${details}.` : ''}`;
  const attempts = Array.isArray(error?.attempts) && error.attempts.length
    ? error.attempts
    : [summarizeFailure(error)];
  if (error?.circuitOpen || /circuit open/i.test(failureText(error))) {
    return formatCircuitWake({ role, details });
  }
  if (cls === "degeneration") {
    return formatDegenerationWake({
      role: role === "review" ? "implementer" : role,
      action: summarizeFailure(error),
      worktree,
      details,
    });
  }
  if (cls === "permanent" || (cls === "process" && !error?.exhausted)) {
    return formatPermanentWake({ role, summary: summarizeFailure(error), details });
  }
  return formatSpentProviderWake({ role, attempts, details, headSha, worktree });
}

export function normalizeLoopValue(value) {
  if (value == null) return "";
  if (typeof value === "string") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) return value.map(normalizeLoopValue).join(",");
  if (typeof value === "object") {
    const keys = Object.keys(value).sort();
    return keys.map((key) => `${key}=${normalizeLoopValue(value[key])}`).join("&");
  }
  return String(value);
}

export function createLoopGuard({
  history = LOOP_HISTORY,
  warnAt = LOOP_WARN_AT,
  stopAt = LOOP_STOP_AT,
} = {}) {
  const events = [];
  let warned = false;
  function fingerprint({ name, args, result, state }) {
    return `${name}|${normalizeLoopValue(args)}|${normalizeLoopValue(result)}|${normalizeLoopValue(state)}`;
  }
  function countTail(fp) {
    let n = 0;
    for (let i = events.length - 1; i >= 0; i--) {
      if (events[i] === fp) n += 1;
      else break;
    }
    return n;
  }
  function countCycle(period) {
    if (events.length < period * 2) return 0;
    const cycle = events.slice(-period).join("||");
    let n = 0;
    for (let start = events.length - period; start >= 0; start -= period) {
      const slice = events.slice(start, start + period).join("||");
      if (slice !== cycle) break;
      n += 1;
    }
    return n;
  }
  return {
    observe(event) {
      const fp = fingerprint(event);
      events.push(fp);
      if (events.length > history) events.shift();
      const same = countTail(fp);
      const cycle = Math.max(countCycle(2), countCycle(3));
      const occurrences = Math.max(same, cycle);
      if (occurrences >= stopAt) return { action: "stop", occurrences, warned };
      if (occurrences >= warnAt) {
        warned = true;
        return { action: "warn", occurrences, warned };
      }
      return { action: "allow", occurrences, warned };
    },
    get warned() {
      return warned;
    },
    get events() {
      return [...events];
    },
  };
}

export const LOOP_WARNING =
  "This action pattern has repeated three times with unchanged results. Use the existing result, change the investigation, or finish. Continuing the same pattern will stop this run.";

File: runtimes/python/src/shared/recovery.py
import json
import random
import re
import sys
import time
import dataclasses
from enum import Enum

PROVIDER_ATTEMPTS = 3
MALFORMED_REGENS = 1
BACKOFF_CAPS = (5.0, 10.0)

RESULT_PREFIX = "RESEARCHER_RESULT "
HEARTBEAT_LINE = "RESEARCHER_HEARTBEAT"


class DegenerationError(RuntimeError):
    def __init__(self, action="repeated tool calls", occurrences=6):
        super().__init__(
            f"repeated-action loop persisted despite a warning: {action}, {occurrences} occurrences"
        )
        self.action = action
        self.occurrences = occurrences
        self.failure_class = "degeneration"


class RecoveryExhausted(RuntimeError):
    def __init__(self, attempts):
        chain = " → ".join(attempts) if attempts else "provider failure"
        super().__init__(f"provider recovery exhausted: {chain}")
        self.attempts = list(attempts)
        self.failure_class = "transient"


class PermanentError(RuntimeError):
    failure_class = "permanent"


def cause_chain(error):
    seen = set()
    while error is not None and id(error) not in seen:
        seen.add(id(error))
        yield error
        error = error.__cause__ or error.__context__


def failure_text(error):
    return " | ".join(str(item) for item in cause_chain(error)) if isinstance(error, BaseException) else str(error or "")


def failure_details(error):
    chain = list(cause_chain(error))
    typed = next((item for item in chain if getattr(item, 'failure_class', None)), error)
    response = next((getattr(item, 'response', None) for item in chain if getattr(item, 'response', None) is not None), None)
    return {
        'message': failure_text(error), 'failureClass': classify_failure(error),
        'attempts': getattr(typed, 'attempts', []), 'exhausted': getattr(typed, 'exhausted', isinstance(typed, RecoveryExhausted)),
        'retryAfter': response.headers.get('Retry-After') if response is not None else None,
    }



def classify_failure(error):
    for item in cause_chain(error) if isinstance(error, BaseException) else []:
        if getattr(item, 'failure_class', None): return item.failure_class
        status = getattr(item, 'status_code', None)
        if status in (400, 401, 403, 404, 422): return 'permanent'
        if status in (408, 429, 500, 502, 503, 504): return 'transient'
    s = failure_text(error).lower()
    if "degeneration" in s or "repeated-action loop" in s:
        return "degeneration"
    if re.search(
        r"set brave_api_key|exa_api_key|401\b|403\b|\b400\b|invalid api key|authentication|"
        r"insufficient_quota|quota exceeded|unsupported model|invalid_request|"
        r"invalid-argument|invalid_argument|badrequesterror|bad request|"
        r"maximum prompt|context length|context window|too many tokens|token limit",
        s,
    ) and not re.search(r"timeout|502|503|429", s):
        return "permanent"
    if re.search(
        r"\b429\b|rate.?limit|\b50[0234]\b|econnreset|etimedout|readtimeout|"
        r"timeout after|internal error during token|"
        r"service unavailable|temporarily unavailable",
        s,
    ):
        return "transient"
    if re.search(r"empty model response|produced no output|question is empty", s):
        return "invalid_output"
    return "process"


def summarize_failure(error):
    text = re.sub(r"\s+", " ", failure_text(error)).strip()
    timeout = re.search(r"timeout after ([\d.]+)\s*seconds", text, re.I)
    if timeout:
        return f"ReadTimeout at the provider's {timeout.group(1)}-second timeout"
    xai = re.search(r'"error"\s*:\s*"((?:\\.|[^"\\])*)"', text)
    if xai:
        return xai.group(1).replace('\\"', '"')
    if re.search(r"internal error during token|\b500\b", text, re.I) and not re.search(r"\b400\b", text):
        return "HTTP 500"
    if re.search(r"readtimeout", text, re.I):
        return "ReadTimeout"
    http = re.search(r"\bHTTP\s*([45]\d\d)\b", text, re.I)
    if http:
        return f"HTTP {http.group(1)}"
    return text if len(text) <= 240 else text[-240:]


def backoff_seconds(failed_attempt, rng=random.random):
    cap = BACKOFF_CAPS[min(max(failed_attempt, 1), len(BACKOFF_CAPS)) - 1]
    return rng() * cap


def emit_result(payload, stream=None):
    target = sys.stderr if stream is None else stream
    print(f"{RESULT_PREFIX}{json.dumps(payload, ensure_ascii=False)}", file=target, flush=True)


def empty_model_response(message):
    content = getattr(message, "content", None)
    tool_calls = getattr(message, "tool_calls", None)
    if tool_calls:
        return False
    if content is None:
        return True
    if isinstance(content, str):
        return not content.strip()
    return False


class RetryingModel:
    def __init__(self, inner, sleep=time.sleep, rng=random.random, supervisor=None):
        self._inner = inner
        self._sleep = sleep
        self._rng = rng
        self.attempts = []
        self.regenerations = 0
        self.recovery_actions = 0
        self.supervisor = supervisor
        self.sequence = 0

    def __getattr__(self, name):
        return getattr(self._inner, name)

    def generate(self, *args, **kwargs):
        self.sequence += 1
        if "tools_to_call_from" in kwargs:
            kwargs["tools_to_call_from"] = [tool for tool in kwargs["tools_to_call_from"] if tool.name != "final_answer"]
        if self.supervisor:
            def run():
                result = self._inner.generate(*args, **kwargs)
                if empty_model_response(result): raise RuntimeError('empty model response')
                if isinstance(result.content, str): self.supervisor.call('text', f'model:{self.sequence}', {'text': result.content})
                return result
            # Save actual messages and tool schemas without credential-bearing model configuration.
            def serial(value):
                # Journal the schema already loaded in memory. Tool.to_dict()
                # exports executable code and reparses mutable source files.
                if hasattr(value, 'inputs') and hasattr(value, 'name'):
                    return {'name': value.name, 'description': value.description, 'inputs': serial(value.inputs), 'output_type': value.output_type}
                if hasattr(value, 'model_dump'): return serial(value.model_dump(mode='json'))
                if dataclasses.is_dataclass(value): return serial(dataclasses.asdict(value))
                if isinstance(value, Enum): return value.value
                if hasattr(value, 'to_dict'): return serial(value.to_dict())
                if isinstance(value, dict): return {k: serial(v) for k, v in value.items()}
                if isinstance(value, (list, tuple)): return [serial(v) for v in value]
                return value
            request = {'model': self._inner.model_id, 'messages': serial(args), 'options': serial(kwargs)}
            def repair(feedback):
                from smolagents.models import ChatMessage
                messages = kwargs.get('messages', args[0] if args else None)
                if isinstance(messages, list): messages.append(ChatMessage(role='user', content=feedback))
                request.update(messages=serial(args), options=serial(kwargs))
            run.repair = repair
            return self.supervisor.request(f'model:{self.sequence}', request, run, serial)
        self.attempts = []
        last = None
        for n in range(1, PROVIDER_ATTEMPTS + 1):
            try:
                result = self._inner.generate(*args, **kwargs)
                if empty_model_response(result):
                    raise RuntimeError("empty model response")
                return result
            except DegenerationError:
                raise
            except Exception as error:
                cls = classify_failure(error)
                summary = summarize_failure(error)
                self.attempts.append(summary)
                last = error
                if cls == "permanent":
                    raise PermanentError(summary) from error
                if cls == "process":
                    raise
                if cls == "degeneration":
                    raise
                if cls == "invalid_output":
                    if self.regenerations >= MALFORMED_REGENS:
                        raise RecoveryExhausted(self.attempts) from error
                    self.regenerations += 1
                if n >= PROVIDER_ATTEMPTS or self.recovery_actions >= 6:
                    raise RecoveryExhausted(self.attempts) from error
                self.recovery_actions += 1
                self._sleep(backoff_seconds(n, self._rng))
        raise RecoveryExhausted(self.attempts) from last
